pandas/core/groupby.py

```diff
@@ -3742,7 +3742,22 @@ def _reindex_output(self, result):
         levels_list = [ping.group_index for ping in groupings]
         index = MultiIndex.from_product(levels_list, names=self.grouper.names)
         d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+
+        if not self.as_index:
+            result.set_index(self.grouper.result_index, inplace=True)
+
+        result = result.reindex(**d).sortlevel(axis=self.axis)
+
+        if not self.as_index:
+            idx = result.index
```

Comment:
It is kludgey. I'll try to make it cleaner but I'm not sure if I can hide the ugly part. Some the old code (3 lines?) can go to MultiIndex (refactoring Categorical levels) but the new part consists of operation on a data frame.

The outline:
- the old code is:
```python
# group keys are already in result.index
new_index = refactor(old_index)       # expanding categorical levels
result.reindex(new_index)
```
- the new code (naive version):
```python
if self.as_index == False:
# result has a flat index, group keys are in result.columns
    result.set_index(group keys)              # (1)  now, result.index == old_index
    result.reindex(new_index)                 # (2)  the old part
    result.reset_index()                      # (3)
```
(1) and (3) involve, respectively, dropping and inserting columns into `result` and they don't go well with hierachical columns (if `result` has them). So instead, I set index with an equivalent object in (1) and copy index levels to respective columns (if possible) in (3).